File: src/position_motion.cpp
/*!*******************************************************************************************
 *  \file       position_motion.cpp
 *  \brief      This file contains the implementation of the PositionMotion class.
 *  \authors    Miguel Fernández Cortizas
 *              Pedro Arias Pérez
 *              David Pérez Saura
 *              Rafael Pérez Seguí
 *
 *  \copyright  Copyright (c) 2022 Universidad Politécnica de Madrid
 *              All Rights Reserved
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 3. Neither the name of the copyright holder nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ********************************************************************************/

#include "motion_reference_handlers/position_motion.hpp"

namespace as2
{
    namespace motionReferenceHandlers
    {
        PositionMotion::PositionMotion(as2::Node *node_ptr) : BasicMotionReferenceHandler(node_ptr)
        {
            desired_control_mode_.yaw_mode = as2_msgs::msg::ControlMode::NONE;
            desired_control_mode_.control_mode = as2_msgs::msg::ControlMode::POSITION;
            desired_control_mode_.reference_frame = as2_msgs::msg::ControlMode::LOCAL_ENU_FRAME;
        };

        bool PositionMotion::sendPositionCommandWithYawAngle(
            const float &x, const float &y, const float &z, const float &yaw_angle,
            const float &vx = 0.0f, const float &vy = 0.0f, const float &vz = 0.0f)
        {
            return sendPositionCommandWithYawAngle(
                x, y, z, tf2::toMsg(tf2::Quaternion(tf2::Vector3(0, 0, 1), yaw_angle)),
                vx, vy, vz);
        }

        bool PositionMotion::sendPositionCommandWithYawAngle(
            const float &x, const float &y, const float &z, const geometry_msgs::msg::Quaternion &q,
            const float &vx = 0.0f, const float &vy = 0.0f, const float &vz = 0.0f)
        {
            desired_control_mode_.yaw_mode = as2_msgs::msg::ControlMode::YAW_ANGLE;
            this->command_pose_msg_.pose.position.x = x;
            this->command_pose_msg_.pose.position.y = y;
            this->command_pose_msg_.pose.position.z = z;

            this->command_pose_msg_.pose.orientation = q;

            this->command_twist_msg_.twist.linear.x = vx;
            this->command_twist_msg_.twist.linear.y = vy;
            this->command_twist_msg_.twist.linear.z = vz;

            return this->sendCommand();
        };

        bool PositionMotion::sendPositionCommandWithYawSpeed(
            const float &x, const float &y, const float &z, const float &yaw_speed,
            const float &vx = 0.0f, const float &vy = 0.0f, const float &vz = 0.0f)
        {
            desired_control_mode_.yaw_mode = as2_msgs::msg::ControlMode::YAW_SPEED;

            this->command_pose_msg_.pose.position.x = x;
            this->command_pose_msg_.pose.position.y = y;
            this->command_pose_msg_.pose.position.z = z;

            this->command_twist_msg_.twist.angular.z = yaw_speed;

            this->command_twist_msg_.twist.linear.x = vx;
            this->command_twist_msg_.twist.linear.y = vy;
            this->command_twist_msg_.twist.linear.z = vz;

            return this->sendCommand();
        };
    } // namespace motionReferenceHandlers
} // namespace as2

File: motion_reference_handlers/position_motion.py
from motion_reference_handlers.basic_motion_references import BasicMotionReferenceHandler
import motion_reference_handlers.utils as utils
from as2_msgs.msg import ControlMode
from geometry_msgs.msg import PoseStamped, TwistStamped


class PositionMotion(BasicMotionReferenceHandler):
    def __init__(self, node):
        super().__init__(node)
        self.desired_control_mode_.yaw_mode = ControlMode.NONE
        self.desired_control_mode_.control_mode = ControlMode.POSITION
        self.desired_control_mode_.reference_frame = ControlMode.UNDEFINED_FRAME

    def __own_send_command(self, yaw_mode, pose_msg, twist_mgs):
        self.desired_control_mode_.yaw_mode = yaw_mode
        self.command_pose_msg_ = pose_msg
        self.command_twist_msg_ = twist_mgs
        send_pose = self.sendPoseCommand()
        send_twist = self.sendTwistCommand()
        return send_pose and send_twist

    def __check_input_pose(self, pose, pose_frame_id):
        pose_msg = PoseStamped()
        if isinstance(pose, list):
            pose_msg.pose.position.x = float(pose[0])
            pose_msg.pose.position.y = float(pose[1])
            pose_msg.pose.position.z = float(pose[2])
            pose_msg.header.frame_id = pose_frame_id
        elif isinstance(pose, PoseStamped):
            pose_msg = pose
        else:
            self.node.get_logger().error("Pose is not a list or PoseStamped")
            return None

        if pose_msg.header.frame_id == '':
            self.node.get_logger().error("Pose frame id is not set")
            return None
        return pose_msg

    def __check_input_twist(self, twist, twist_frame_id):
        twist_mgs = TwistStamped()

        if isinstance(twist, float):
            twist_mgs.twist.linear.x = twist
            twist_mgs.twist.linear.y = twist
            twist_mgs.twist.linear.z = twist
            twist_mgs.header.frame_id = twist_frame_id
        elif isinstance(twist, list):
            twist_mgs.twist.linear.x = float(twist[0])
            twist_mgs.twist.linear.y = float(twist[1])
            twist_mgs.twist.linear.z = float(twist[2])
            twist_mgs.header.frame_id = twist_frame_id
        elif isinstance(twist, TwistStamped):
            twist_mgs = twist
        elif twist == None:
            # Default value -> no limit
            twist_mgs.header.frame_id = 'earth'
        else:
            self.node.get_logger().error("Twist is not a float, list or TwistStamped")
            return None

        if twist_mgs.header.frame_id == '':
            self.node.get_logger().error("Twist frame id is not set")
            return None

        return twist_mgs

    def send_position_command_with_yaw_angle(self, pose, twist_limit=None, pose_frame_id='', twist_frame_id='', yaw_angle=None):
        pose_msg = self.__check_input_pose(pose, pose_frame_id)
        twist_msg = self.__check_input_twist(twist_limit, twist_frame_id)

        if twist_msg == None or pose_msg == None:
            return False

        if isinstance(pose, list):
            if not isinstance(yaw_angle, float):
                self.node.get_logger().error(
                    "Yaw angle is not set")
                return False
            pose_msg.pose.orientation = utils.get_quaternion_from_yaw_angle(
                yaw_angle)

        return self.__own_send_command(ControlMode.YAW_ANGLE, pose_msg, twist_msg)

    def send_position_command_with_yaw_speed(self, pose, twist_limit=None, pose_frame_id='', twist_frame_id='', yaw_speed=None):
        pose_msg = self.__check_input_pose(pose, pose_frame_id)
        twist_msg = self.__check_input_twist(twist_limit, twist_frame_id)

        if twist_msg == None or pose_msg == None:
            return False

        if isinstance(twist_limit, list):
            if not isinstance(yaw_speed, float):
                self.node.get_logger().error(
                    "Yaw speed is not set")
                return False

        if yaw_speed != None:
            twist_msg.twist.angular.z = yaw_speed

        return self.__own_send_command(ControlMode.YAW_SPEED, pose_msg, twist_msg)
